src/share/poudriere/image.sh

```diff
@@ -130,10 +135,83 @@ mkminiroot() {
 	gzip -9 ${OUTPUTDIR}/${IMAGENAME}-miniroot
 }
 
+create_zfs_be_datasets() {
+	local OPT OPTSTR
+
+	if [ -n "$1" ]; then
+		. $1
+	fi
+	: ${ZFS_DATASETS:="
+		# DATASET	OPTIONS (space separated)
+
+		# Boot Environment [BE] root and default boot dataset
+		/$ZFS_BEROOT_NAME			mountpoint=none
+		/$ZFS_BEROOT_NAME/$ZFS_BOOTFS_NAME	mountpoint=/
+
+		# Compress /tmp, allow exec but not setuid
+		/tmp		mountpoint=/tmp exec=on setuid=off
+
+		# Don't mount /usr so that 'base' files go to the BEROOT
+		/usr		mountpoint=/usr canmount=off
+
+		# Home directories separated so they are common to all BEs
+		/usr/home	# NB: /home is a symlink to /usr/home
+
+		# Ports tree
+		/usr/ports	setuid=off
+
+		# Source tree (compressed)
+		/usr/src
+
+		# Create /var and friends
+		/var		mountpoint=/var canmount=off
+		/var/audit	exec=off setuid=off
+		/var/crash	exec=off setuid=off
+		/var/log	exec=off setuid=off
+		/var/mail	atime=on
+		/var/tmp	setuid=off
+	"}
+
+	msg "Creating ZFS Datasets"
+	echo "$ZFS_DATASETS" | while read dataset options; do
+		# Skip blank lines and comments
+		case "$dataset" in "#"*|"") continue; esac
+		# Remove potential inline comments in options
+		options="${options%%#*}"
+		# Replace tabs with spaces
+		str_replaceall "$options" "	" " " options
+		# Reduce contiguous runs of space to one single space
+		oldoptions=
+		while [ "$oldoptions" != "$options" ]; do
+			oldoptions="$options"
+			str_replaceall "$options" "  " " " options
+		done
+		# Replace both commas and spaces with ` -o '
+		str_replaceall "$options" "[ ,]" " -o " options
+		# Create the dataset with desired options
+		zfs create ${options:+-o $options} "$zroot$dataset" ||
+		    err 1 "Dataset creation failed"
+	done
+}
+
+zfssend_writereplicationstream() {
+	# Arguments:
+	# $1: snapshot to recursively replicate
+	# $2: Image name to write replication stream to
```

Comment:
```
[ $# -eq 2 ] || eargs zfssend_writereplicationstream snapshot_ftom image_to
```
Causes an error if called wrong and is kinda self-documenting. Keep your comments if you want.